servers/src/main/java/tachyon/master/next/IndexedSet.java

```diff
@@ -0,0 +1,225 @@
+  private final Map<FieldIndex<T>, Integer> mIndexMap;
+  /** List of maps from fieldValue to set of T which is indexed by field name */
+  private final List<Map<Object, Set<T>>> mSetIndexedByFieldValue;
+  /** List of {@link Field} which is indexed by field name */
+  private final List<Field> mFields;
+
+  /**
+   * A wrapper around a field name string to force the user to predefine the indexes, the user can
+   * reuse the instance of this class as parameters later other than directly use the error prone
+   * raw Strings.
+   */
+  public static interface FieldIndex<T> {
+    Object getFieldValue(T o);
+  }
+
+  /**
+   * Construct a new IndexedSet with at least one field as the index, the field can be either public
+   * or private. The {@link Field}s for these fields aren't retrieved until {@link #add(Object)} is
+   * called for the first time, it will be cached once it is retrieved so that reflection will only
+   * be needed once for a field.
+   *
+   * @param field at least one field is needed to index the set of objects
+   * @param otherFields other fields to index the set
+   */
+  public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
+    mObjects = new HashSet<T>();
+
+    mIndexMap = new HashMap<FieldIndex<T>, Integer>(otherFields.length + 1);
+    mIndexMap.put(field, 0);
+    for (int i = 1; i <= otherFields.length; i ++) {
+      mIndexMap.put(otherFields[i - 1], i);
+    }
+
+    mFields = new ArrayList<Field>(mIndexMap.size());
+    for (int i = 0; i < mIndexMap.size(); i ++) {
+      mFields.add(null);
+    }
+
+    mSetIndexedByFieldValue = new ArrayList<Map<Object, Set<T>>>(mIndexMap.size());
+    for (int i = 0; i < mIndexMap.size(); i ++) {
+      mSetIndexedByFieldValue.add(new HashMap<Object, Set<T>>());
+    }
+  }
+
+  /**
+   * Add an object o to the set if there is no other object o2 such that (o == null ? o2 == null :
+   * o.equals(o2)). If this set already contains the object, the call leaves the set unchanged.
+   *
+   * @param object the object to add
+   * @return true if the object is successfully added to all indexes, otherwise false
+   */
+  public boolean add(T object) {
+    synchronized (mFields) {
+      boolean success = mObjects.add(object);
+      for (FieldIndex<T> field : mIndexMap.keySet()) {
+        Map<Object, Set<T>> fieldValueToSet = mSetIndexedByFieldValue.get(mIndexMap.get(field));
+        Object value = field.getFieldValue(object);
+        if (fieldValueToSet.containsKey(value)) {
+          success = success && fieldValueToSet.get(value).add(object);
+        } else {
+          fieldValueToSet.put(value, Sets.newHashSet(object));
+        }
+      }
+      return success;
+    }
+  }
+
+  /**
+   * Return the set of all objects, the returned set is backed up by the internal set, any changes
+   * to the internal set are reflected in the returned set, and vice-versa.
+   *
+   * @return a set of all objects
+   */
+  public Set<T> all() {
```

Comment:
It seems all the usage of `all()` are in foreach loop. Can we simply make this `IndexedSet` class implementing `Iterable` to avoid this method then?